File: BMWCenter/Core/Care/BadgeService.swift
import Foundation
import SwiftData

@MainActor
final class BadgeService {
    private let modelContext: ModelContext

    init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }

    func awardedKeys() -> Set<String> {
        let rows = (try? modelContext.fetch(FetchDescriptor<BadgeAward>())) ?? []
        return Set(rows.map(\.badgeKey))
    }

    /// Awards only when vehicle is stopped. Returns new badge keys.
    @discardableResult
    func evaluateAwards(
        isStopped: Bool,
        cleanWarmups: Int,
        compliantCooldowns: Int,
        harshAccelKm: Double,
        tripCount: Int,
        longHaulKm: Double,
        thermostatCaught: Bool
    ) -> [String] {
        guard isStopped else { return [] }
        var newly: [String] = []
        func award(_ key: String, progress: Double) {
            guard !awardedKeys().contains(key) else { return }
            modelContext.insert(BadgeAward(badgeKey: key, progressSnapshot: progress))
            newly.append(key)
        }
        if cleanWarmups >= 25 { award("coldBlooded", progress: Double(cleanWarmups)) }
        if compliantCooldowns >= 50 { award("turboWhisperer", progress: Double(compliantCooldowns)) }
        if harshAccelKm >= 1000 { award("featherFoot", progress: harshAccelKm) }
        if thermostatCaught { award("thermostatSentinel", progress: 1) }
        if tripCount >= 100 { award("century", progress: Double(tripCount)) }
        if longHaulKm >= 300 { award("longHauler", progress: longHaulKm) }
        if !newly.isEmpty { try? modelContext.save() }
        return newly
    }

    static func titleKey(for badgeKey: String) -> String {
        switch badgeKey {
        case "coldBlooded": return "badge.coldBlooded"
        case "turboWhisperer": return "badge.turboWhisperer"
        case "featherFoot": return "badge.featherFoot"
        case "thermostatSentinel": return "badge.thermostatSentinel"
        case "century": return "badge.century"
        case "frugal": return "badge.frugal"
        case "longHauler": return "badge.longHauler"
        default: return "badge.\(badgeKey)"
        }
    }
}

File: mobile/src/core/care/badgeService.ts
import { db } from "../storage/db";
import { badgeAwards } from "../storage/schema";

export function badgeTitleKey(badgeKey: string): string {
  const known = ["coldBlooded", "turboWhisperer", "featherFoot", "thermostatSentinel", "century", "frugal", "longHauler"];
  return known.includes(badgeKey) ? `badge.${badgeKey}` : `badge.${badgeKey}`;
}

class BadgeService {
  async awardedKeys(): Promise<Set<string>> {
    const rows = await db.select().from(badgeAwards);
    return new Set(rows.map((r) => r.badgeKey));
  }

  /** Awards only while the vehicle is stopped. Returns newly-earned badge keys. */
  async evaluateAwards(params: {
    isStopped: boolean;
    cleanWarmups: number;
    compliantCooldowns: number;
    harshAccelKm: number;
    tripCount: number;
    longHaulKm: number;
    thermostatCaught: boolean;
  }): Promise<string[]> {
    if (!params.isStopped) return [];
    const awarded = await this.awardedKeys();
    const newly: string[] = [];
    const award = async (key: string, progress: number) => {
      if (awarded.has(key)) return;
      await db.insert(badgeAwards).values({ badgeKey: key, awardedAt: Date.now(), progressSnapshot: progress });
      newly.push(key);
    };
    if (params.cleanWarmups >= 25) await award("coldBlooded", params.cleanWarmups);
    if (params.compliantCooldowns >= 50) await award("turboWhisperer", params.compliantCooldowns);
    if (params.harshAccelKm >= 1000) await award("featherFoot", params.harshAccelKm);
    if (params.thermostatCaught) await award("thermostatSentinel", 1);
    if (params.tripCount >= 100) await award("century", params.tripCount);
    if (params.longHaulKm >= 300) await award("longHauler", params.longHaulKm);
    return newly;
  }
}

export const badgeService = new BadgeService();
